keps/sig-apps/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1087 @@
+  the `spec.resources.requests.storage` of the template; and
+- `status.currentVolumeAttributesClassName` of PVC is equal to
+  the `spec.volumeAttributesClassName` of the template.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1: Batch Expand Volumes
+
+We're running a CI/CD system and the end-to-end automation is desired.
+To expand the volumes managed by a StatefulSet,
+we can just use the same pipeline that we are already using to update the Pod.
+All the test, review, approval, and rollback process can be reused.
+
+#### Story 2: Shinking the PV by Re-creating PVC
+
+After running our app for a while, we optimize the data layout and reduce the required storage size.
+Now we want to shrink the PVs to save cost.
+We can not afford any downtime, so we don't want to delete and recreate the StatefulSet.
+We also don't have the infrastructure to migrate between two StatefulSets.
+Our app can automatically rebuild the data in the new storage from other replicas.
+So we update the `volumeClaimTemplates` of the StatefulSet,
+delete the PVC and Pod of one replica, let the controller re-create them,
+then monitor the rebuild process.
+Once the rebuild completes successfully, we proceed to the next replica.
+
+#### Story 3: Asymmetric Replicas
+
+The storage requirement of different replicas are not identical,
+so we still want to update each PVC manually and separately.
+Possibly we also update the `volumeClaimTemplates` for new replicas,
+but we don't want the controller to interfere with the existing replicas.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+When designing the `InPlace` update strategy, we update the PVC like how we re-create the Pod.
+i.e. we update the PVC whenever we would re-create the Pod;
+we wait for the PVC to be compatible whenever we would wait for the Pod to be available.
+
+The StatefulSet controller should also keeps the current and updated revision of the `volumeClaimTemplates`,
+so that a StatefulSet can still re-create Pods and PVCs that are yet-to-be-updated.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+TODO: Recover from failed in-place update (insufficient storage, etc.)
```

Comment:
This section will need to cover all the possible edge cases, and failure scenarios, since I believe there will be a few. The tables I requested covering all the possible cases should help you identify those.